File: 브루트포스/9663.py
import sys
input = sys.stdin.readline

def calc(row):
    if row == n:
        return 1
    ans = 0
    for i in range(n):
        if check(row,i):
            check_dig[row+i] = True
            check_dig2[row-i+n-1] = True
            check_col[i] = True
            a[row][i] = True
            ans+=calc(row+1)
            check_dig[row+i] = False
            check_dig2[row-i+n-1] = False
            check_col[i] = False
            a[row][i] = False
    return ans

def check(row, col):
    if check_col[col]:
        return False
    if check_dig[row+col]:
        return False
    if check_dig2[row-col+n-1]:
        return False
    return True


n = int(input().rstrip())
a = [[False]*n for _ in range(n)]
check_col = [False]*n # i번 열에 퀸이 놓여져 있는지 여부
check_dig = [False]*(2*n-1) # / 방향 대각선에 퀸이 있는지 여부
check_dig2 = [False]*(2*n-1) # \ 방향 대각선에 퀸이 있는지 여부
print(calc(0))

File: 브루트포스/9663.java
import java.util.*;

public class Main {

    public static int n;
    public static boolean[][] a = new boolean[15][15];
    public static boolean[] check_col = new boolean[15]; // i번 열에 퀸이 놓여 있는지 여부
    public static boolean[] check_dig = new boolean[40]; // / 방향 대각선에 퀸이 있는지 여부
    public static boolean[] check_dig2 = new boolean[40]; // \ 방향 대각선에 퀸이 있는지 여부

    public static boolean check(int row, int col) {
        // |
        if (check_col[col]) {
            return false;
        }
        // 왼쪽 위 대각선
        if (check_dig[row+col]) {
            return false;
        }
        // /
        if (check_dig2[row-col+n]) {
            return false;
        }
        return true;
    }
    public static int calc(int row) {
        if (row == n) {
            // ans += 1;
            return 1;
        }
        int cnt = 0;
        for (int col=0; col<n; col++) {
            if (check(row, col)) {
                check_dig[row+col] = true;
                check_dig2[row-col+n] = true;
                check_col[col] = true;
                a[row][col] = true;
                cnt += calc(row+1);
                check_dig[row+col] = false;
                check_dig2[row-col+n] = false;
                check_col[col] = false;
                a[row][col] = false;
            }
        }
        return cnt;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        System.out.println(calc(0));
    }
}
